[Validator] Fix traverse option on Valid constraint when used as Attribute

tobias-93 opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #46243
| License       | MIT
| Doc PR        | n.a.

Fix Valid constraint when using as Attribute.

tobias-93 committed:
[Validator] Fix traverse option on Valid constraint when used as Attribute

nicolas-grekas commented:
Can you please add a test case? Note also that we cannot use mixed as 5.4 should support PHP 7.2

tobias-93 commented:
@nicolas-grekas Thanks for the feedback. I have removed the `mixed` parameter, and added a test. The remaining failing tests are parts of Symfony I didn't touch, so that's hopefully not my fault...

xabbuh commented on src/Symfony/Component/Validator/Tests/Constraints/ValidTest.php:
```diff
@@ -32,4 +34,34 @@ public function testGroupsAreNullByDefault()
+    #[Valid]
+    private $a;
+
+    #[Valid(groups: ['traverse_group'], traverse: false)] // Needs a group to work at all for this test
```
Why is the group needed?

tobias-93 replied:
See https://github.com/symfony/symfony/blob/2d44f229ff9cb169c4c90f2c05e60069350eb119/src/Symfony/Component/Validator/Mapping/GenericMetadata.php#L145, if `$constraint->groups === null` the constraint itself is not added but the traversalstrategy is changed. Since we want to check if the argument properly propagates to the constraint (and not the traversalstrategy or other functional behavior) I added a group to demonstrate.

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/Valid.php:
```diff
@@ -43,4 +43,11 @@ public function addImplicitGroupName(string $group)
         }
     }
+
+    public function __construct(array $options = null, bool $traverse = null, array $groups = null, $payload = null)
```
shouldn't $traverse be the last argument to not break BC when creating an instance?

tobias-93 replied:
I did it similarly to the constraints that were fixed in https://github.com/symfony/symfony/commit/fb99eb2052523cfa2a9ba36006d020796548aaa6, in those argument order wasn't honored either. Why would that be an issue for this constraint?

nicolas-grekas replied:
/cc @derrabus I may miss something here

fabpot replied:
I think it does not hurt to not break BC anyway, right? Let's reorder.
Also, the `__construct` method should be moved as the first method of the class.

tobias-93 replied:
@fabpot reordered, thanks for the remark. I actually just ran into an issue where the argument order was not honored (the `Callback` constraint) so I think keeping the BC is the right decision 👍🏼

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/Valid.php:
```diff
@@ -43,4 +43,11 @@ public function addImplicitGroupName(string $group)
+    {
+        parent::__construct($options ?? [], $groups, $payload);
+
+        $this->traverse = $traverse ?? $this->traverse ?? true;
```
> `$this->traverse ?? true`

can `$this->traverse` really be null?

HeahDude replied:
What about `bool $traverse = true` in the constructor instead? or using `?bool` there

tobias-93 replied:
Fallback to true removed to bring it in line with https://github.com/symfony/symfony/commit/fb99eb2052523cfa2a9ba36006d020796548aaa6

tobias-93 commented:
@nicolas-grekas @derrabus Any progress here?

HeahDude commented on src/Symfony/Component/Validator/Constraints/Valid.php:
```diff
@@ -43,4 +43,11 @@ public function addImplicitGroupName(string $group)
+    {
+        parent::__construct($options ?? [], $groups, $payload);
+
+        $this->traverse = $traverse ?? $this->traverse;
```
I don't see the point of such check, what about `bool $traverse = true` and a simple assignation here?

stof replied:
`traverse` might be assigned through `$options` too

fabpot commented:
Thank you @tobias-93.
